### PR TITLE
build.d: Display dependency informations only on failure

### PR DESCRIPTION
**Pull request by MoonlightSentinel**
This PR tries to improve `build.d`s output by leveraging the changes introduced in #10560. `build.d` will now print a summary for any failed `Dependency` (and its parents), but only the specified message on success. This should achieve a better signal/noise ratio while also providing enough information to track down the cause of the build failure.

Furthermore this should improve the performance of `build.d` by avoiding unnecessary string formatting, hence producing less garbage.

The first commit enables the advanced output while the second commit removes preceding output/formatting from the declared `Dependency`s.

Example (Compilation error in `aggregate.d`):

```
(TX) VERSION
(TX) DMD_CONF
(TX) SYSCONFDIR
(DC) D_BACKEND_OBJS
(DC) D_LEXER_OBJ
(LIB) BACKEND.lib
(DC) DMD
Dependencies:
 -> D:\repos\Dlang\forks\dmd\generated\windows\release\64\VERSION
 -> D:\repos\Dlang\forks\dmd\generated\windows\release\64\SYSCONFDIR.imp
 -> lexer
 -> backend

Sources:
 -> D:\repos\Dlang\forks\dmd\src\dmd\access.d
 -> D:\repos\Dlang\forks\dmd\src\dmd\aggregate.d
...

Targets:
 -> D:\repos\Dlang\forks\dmd\generated\windows\release\64\dmd.exe

Command: C:\D\dmd2\windows\bin\dmd.exe -ofD:\repos\Dlang\forks\dmd\generated\windows\release\64\dmd.exe -vtls -JD:\repos\Dlang\forks\dmd\res -L/STACK:8388608 -version=MARS -w -de  -m64 -JD:\repos\Dlang\forks\dmd\generated\windows\release\64 -dip25 -g dmd\access.d dmd\aggregate.d dmd\aliasthis.d dmd\apply.d dmd\argtypes.d dmd\argtypes_sysv_x64.d dmd\arrayop.d dmd\arraytypes.d dmd\ast_node.d dmd\astbase.d dmd\astcodegen.d dmd\attrib.d dmd\blockexit.d dmd\builtin.d dmd\canthrow.d dmd\cli.d dmd\clone.d dmd\compiler.d dmd\complex.d dmd\cond.d dmd\constfold.d dmd\cppmangle.d dmd\cppmanglewin.d dmd\ctfeexpr.d dmd\ctorflow.d dmd\dcast.d dmd\dclass.d dmd\declaration.d dmd\delegatize.d dmd\denum.d dmd\dimport.d dmd\dinifile.d dmd\dinterpret.d dmd\dmacro.d dmd\dmangle.d dmd\dmodule.d dmd\doc.d dmd\dscope.d dmd\dstruct.d dmd\dsymbol.d dmd\dsymbolsem.d dmd\dtemplate.d dmd\dversion.d dmd\env.d dmd\escape.d dmd\expression.d dmd\expressionsem.d dmd\func.d dmd\hdrgen.d dmd\impcnvtab.d dmd\imphint.d dmd\init.d dmd\initsem.d dmd\inline.d dmd\inlinecost.d dmd\intrange.d dmd\json.d dmd\lambdacomp.d dmd\lib.d dmd\libelf.d dmd\libmach.d dmd\libmscoff.d dmd\libomf.d dmd\link.d dmd\mars.d dmd\mtype.d dmd\nogc.d dmd\nspace.d dmd\objc.d dmd\opover.d dmd\optimize.d dmd\parse.d dmd\parsetimevisitor.d dmd\permissivevisitor.d dmd\printast.d dmd\safe.d dmd\sapply.d dmd\scanelf.d dmd\scanmach.d dmd\scanmscoff.d dmd\scanomf.d dmd\semantic2.d dmd\semantic3.d dmd\sideeffect.d dmd\statement.d dmd\statement_rewrite_walker.d dmd\statementsem.d dmd\staticassert.d dmd\staticcond.d dmd\strictvisitor.d dmd\target.d dmd\templateparamsem.d dmd\traits.d dmd\transitivevisitor.d dmd\typesem.d dmd\typinf.d dmd\utils.d dmd\visitor.d dmd\foreachvar.d dmd\irstate.d dmd\toctype.d dmd\glue.d dmd\gluelayer.d dmd\todt.d dmd\tocsym.d dmd\toir.d dmd\dmsc.d dmd\tocvdebug.d dmd\s2ir.d dmd\toobj.d dmd\e2ir.d dmd\eh.d dmd\iasm.d dmd\iasmdmd.d dmd\iasmgcc.d dmd\objc_glue.d dmd\backend\cc.d dmd\backend\cdef.d dmd\backend\cgcv.d dmd\backend\code.d dmd\backend\cv4.d dmd\backend\dt.d dmd\backend\el.d dmd\backend\global.d dmd\backend\obj.d dmd\backend\oper.d dmd\backend\outbuf.d dmd\backend\rtlsym.d dmd\backend\code_x86.d dmd\backend\iasm.d dmd\backend\codebuilder.d dmd\backend\ty.d dmd\backend\type.d dmd\backend\exh.d dmd\backend\mach.d dmd\backend\mscoff.d dmd\backend\dwarf.d dmd\backend\dwarf2.d dmd\backend\xmm.d dmd\backend\dlist.d dmd\backend\melf.d dmd\backend\varstats.di dmd\root\aav.d dmd\root\longdouble.d dmd\root\man.d dmd\root\response.d dmd\root\speller.d dmd\root\string.d dmd\root\strtold.d ..\generated\windows\release\64\lexer.o ..\generated\windows\release\64\backend.o
```
Let me know what you think @wilzbach, @marler8997

### CODE REVIEW

**Comment by marler8997** [comment]
There's some benefit to displaying more information during a normal build such as the files being compiled.  It gives us more information in the logs to compare between versions, it's useful to compare information between runs even if they didn't fail.

That extra information may or may not be worth the extra characters written to the console, but we should consider this benefit.

**Comment by marler8997** [comment]
There's also the consideration of printing "too much" to the console.  For example, when most tools fail they're usually good about printing the error that occurred, however, with this the error can get lost in a sea of information about the build information that usually wont' be relevant unless you're trying to debug the build itself.  I think most times you're only interested in the output of the tool that failed.

**Comment by marler8997** [comment]
This problem of printing too much information can be easily solved by only printing this in a `verbose` build.

File: src/build.d [hunk 1]
```diff
@@ -1444,6 +1449,30 @@ class Dependency
...
+        {
+            static if (is(T : bool))
+                bool print = what;
+            else
+                bool print = what.length != 0;
+
+            if (print)
+                writer.formattedWrite(fmt, what);
+        }
+
```
Comment (marler8997) [inline src/build.d]: Maybe print a delimiter to distinguish between operations
```D
write("--------------------------------------------------------------------------------\n");
write("This Operation Failed:\n");
```

File: src/build.d [hunk 2]
```diff
@@ -1444,6 +1449,30 @@ class Dependency
             }
         }
     }
+
+    /// Writes relevant informations about this dependency to stdout
+    private void dump()
+    {
+        scope writer = stdout.lockingTextWriter;
```
Comment (marler8997) [inline src/build.d]: For another PR, could use `lockingTextWriter` to make sure the `dryRun` output isn't mangled between multiple threads.

File: src/build.d (hunk 1 above)
Comment (MoonlightSentinel) [reply src/build.d]: Added

**Comment by MoonlightSentinel** [comment]
> There's some benefit to displaying more information during a normal build such as the files being compiled. It gives us more information in the logs to compare between versions, it's sometimes useful to be able to compare information between runs even if they didn't fail.
> 
> That extra information may or may not be worth the extra characters written to the console, but we should consider this benefit.

Thats an interesting point, maybe we should discuss this for every dependency instead. For example i dont see much benefit in printing `sources.lexer` as part of `lexer`s message because it's an explicit list in `build.d` but dynamically selected files may be useful (e.g. `allSources` for `checkwhitespace`)



> There's also the consideration of printing "too much" to the console. For example, when most tools fail they're usually good about printing the error that occurred, however, with this the error can get lost in a sea of information about the build information that usually wont' be relevant unless you're trying to debug the build itself. I think most times you're only interested in the output of the tool that failed.
> 
> EDIT: After playing with this branch I see you print the pertinent error at the bottom which does help this issue. That being said, I think 90% of the time the build data isn't relevant so it still might make sense to put it behing the verbose flag?

The full dependency trace is now limited to the `verbose` mode. Additionally we could only `dump` the currently built dependency on failure (see the last commit)

File: src/build.d [hunk 3]
```diff
@@ -430,24 +430,28 @@ alias toolsRepo = makeDep!((builder, dep) => builder
...
-            chunkLength = 80; // avoid command-line limit on win32
-        foreach (nextSources; allSources.chunks(chunkLength).parallel(1))
-        {
-            const nextCommand = cmdPrefix ~ nextSources;
-            writeln(nextCommand.join(" "));
-            run(nextCommand);
-        }
-    })
-);
+alias checkwhitespace = makeDep!((builder, dep) {
```
Comment (marler8997) [inline src/build.d]: I'd put this in another PR.

**Review by wilzbach — APPROVED** [review]
Generally LGTM. Nice work!

File: src/build.d (hunk 3 above)
Comment (MoonlightSentinel) [reply src/build.d]: Removed

File: src/build.d (hunk 2 above)
Comment (MoonlightSentinel) [reply src/build.d]: See #10622

**Pull request merged**